### ISSUE TITLE
[A11y][iOS]: SnackBar shows more Voice Control labels than expected

### ISSUE THREAD

**Issue by ChristianEdwardPadilla** [flutter/flutter#157630]
Tracking b/325610878.

### Steps to reproduce

1. Enable iOS Voice Control in the settings app.
2. In a Flutter app (minimal sample app below), cause a Material SnackBar widget to be displayed.
3. Observe the Voice Control labels.

### Expected results

There should be labels for interactive UI elements.

### Actual results

There are more labels than there are interactive UI elements.

This was observed for Material2 and Material3 themes.

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(useMaterial3: false),
      title: 'Flutter Demo',
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  void displaySnackBar(BuildContext context) {
    ScaffoldMessenger.of(context).clearSnackBars();
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(
        content: Text('test bar'),
        // // Enable these in order to see more numbered labels with iOS Voice Control.
        // action: SnackBarAction(
        //   label: 'close',
        //   onPressed: () => print('close'),
        // ),
        // showCloseIcon: true,
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title, style: TextStyle(fontFamily: 'ProductSans')),
      ),
      body: Center(
        child: Text(
          'test',
          key: Key('CountText'),
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () => displaySnackBar(context),
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ),
    );
  }
}
```

</details>

### Screenshots or Video

<details open>
<summary>screenshot of extra labels</summary>

![IMG_0093](https://github.com/user-attachments/assets/4d13678d-c277-42b1-911b-d8fd9d3904ea)

</details>

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel google3, on macOS 14.7 23H124 darwin-x64, locale en)
    • Framework revision 3cd12f2c06 (0 days ago), 2024-10-25T00:00:00.000
    • Engine revision c4b0184c87
    • Dart version 3067d697ae

[✓] Android toolchain - develop for Android devices (Android SDK version Stable)
    • Android SDK at google3
    • Platform Stable, build-tools Stable
    • Java binary at: /Applications/Android Studio with Blaze
      Beta.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 21.0.3+-79915915-b509.11)

[✓] VS Code (version 1.94.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.86.0

[✓] Xcode - develop for iOS and macOS (Xcode 15.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer

[✓] Connected device (1 available)
    ! Error: Browsing on the local area network for CP’s iPhone (2). Ensure the
      device is unlocked and attached with a cable or associated with the same
      local area network as this Mac.
      The device must be opted into Developer Mode to connect wirelessly. (code
      -27)
```

</details>

**Comment by huycozy** [comment]
Reproduced this on Flutter stable and master channels. 

<details>
<summary>Flutter versions (stable and master)</summary>

```console
[✓] Flutter (Channel stable, 3.24.4, on macOS 15.0.1 24A348 darwin-x64, locale en-VN)
    • Flutter version 3.24.4 on channel stable at /Users/huynq/Documents/GitHub/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 603104015d (12 hours ago), 2024-10-24 08:01:25 -0700
    • Engine revision db49896cf2
    • Dart version 3.5.4
    • DevTools version 2.37.3
```

```console
[!] Flutter (Channel master, 3.27.0-1.0.pre.231, on macOS 15.0.1 24A348 darwin-x64, locale en-VN)
    • Flutter version 3.27.0-1.0.pre.231 on channel master at /Users/huynq/Documents/GitHub/flutter_master
    ! Warning: `flutter` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 41b362b12c (22 minutes ago), 2024-10-24 22:59:25 -0400
    • Engine revision b1c2ba8c4d
    • Dart version 3.7.0 (build 3.7.0-65.0.dev)
    • DevTools version 2.40.1
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.
```
</details>

**Comment by jmagman** [comment]
@huycozy does this reproduce on Android TalkBack?

**Comment by huycozy** [comment]
>does this reproduce on Android TalkBack?

@jmagman I tried finding the equivalent of iOS's Voice Control on Android TalkBack but couldn't find the exact one. I also used [Accessibility Scanner app](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor) (based on one issue I tried before at https://github.com/flutter/flutter/issues/104908), I encountered #104908 that I think unrelated to this issue. Please let me know if I'm missing something and I will check the issue again on Android.

**Comment by chunhtai** [comment]
I wonder if this is due to the snackbar has a dismiss action.

**Comment by franciscojma86** [comment]
b/325610878

**Comment by huycozy** [comment]
Re-visiting and confirming this on Android with [Voice Access](https://support.google.com/accessibility/android/answer/6151854): The issue is not reproduced on Android, there is only one number shown on FAB when SnackBar opens.

**Comment by jmagman** [comment]
> I wonder if this is due to the snackbar has a dismiss action.

@chunhtai that would cause this to reproduce on all platforms, right? Do you have any idea why this would only be happening on iOS? Based on that we may know if https://github.com/flutter/flutter/issues/158477 is a duplicate.

**Comment by loic-sharma** [comment]
This appears to be a different bug than https://github.com/flutter/flutter/issues/158477.

In https://github.com/flutter/flutter/issues/158477, we don't provide iOS enough information about an accessibility element. As a result, iOS appears to assume the accessibility element responds to user interaction. To fix that bug, it looks like we'll need to provide more information on affected accessibility elements.

However, for this bug it looks like Flutter sends an incorrect semantics tree to the iOS embedder:

![Image](https://github.com/user-attachments/assets/d7d040fb-f0e2-4c1a-9803-f5c2bab6fcd3)

_SemanticsNode#15 is the snackbar. SemanticsNode#16 is the the snackbar's close button._

The snackbar has actions like `tap` and flags like `isButton`. Thus, Voice Control shows a number as a user can interact with a tappable button. I suspect to fix this we'll need to update the framework to generate a different semantics tree. cc @chunhtai.

**Comment by chunhtai** [comment]
I believe this is outside of our control, and it seems to be a ui design issue. We can't drop or merge user defined buttons in snackbar into one node. so I think this is WAI.

Also, it looks like the snackbar is already tap to dismiss, why adding an additional button to dismiss it? A possible workaround is just add an icon without ontap, it will be functionally the same.

Also this seems to be different from the internal issue https://b.corp.google.com/issues/325610878

**Comment by meredithslota** [comment]
Following up, since the internal bug in question is our, ha. 

@chunhtai -- I investigated the mechanics of this in the engine and framework (with some agentic assistance), and I think there may be a slight misunderstanding of what is causing the number to appear:
1. **SnackBar is not tap-to-dismiss:** A standard `SnackBar` content body has no `onTap` or tap-to-dismiss behavior. If a user actually touches the body text or uses Voice Control to say "Tap <number>" on that badge, nothing happens (it fails silently because `SemanticsObject.accessibilityActivate` returns `NO` when there is no `kTap` action).
2. **Why Voice Control numbers it:** In `SemanticsObject.mm`, `accessibilityRespondsToUserInteraction` returns `YES` whenever `node.actions & ~kSystemActions != 0`. Because `SnackBar` includes `onDismiss` for the VoiceOver escape gesture, the node has `SemanticsAction::kDismiss`. Since `kDismiss` is not in `kSystemActions`, Flutter tells iOS that the static container responds to user interaction, prompting Voice Control to place a number badge on an un-tappable element.
3. **The close button is not redundant:** Because the body text is not tap-to-dismiss, the close icon is the only direct-touch affordance the user has to dismiss the bar. We don't want to merge or drop any buttons; we simply need the static container to not claim it is an interactive control.

According to Apple's `UIAccessibility.h`, `accessibilityRespondsToUserInteraction` is meant for elements that perform an action on touch/tap (buttons, sliders, etc.), whereas `kDismiss` maps exclusively to `accessibilityPerformEscape` (the VoiceOver two-finger scrub / Esc key).

Excluding `kDismiss` (and `kShowOnScreen`) from `accessibilityRespondsToUserInteraction` in `SemanticsObject.mm` resolves the issue, I think:
- Voice Control stops placing a dead number badge over the static message.
- The close icon retains its number badge and responds to "Tap <number>".
- VoiceOver two-finger scrub dismiss continues to work unchanged via `accessibilityPerformEscape`. 

Does this seem reasonable?

**Comment by chunhtai** [comment]
reassign to the ios team as the the fix will be in SemanticsObject.mm
